### PR TITLE
KAFKA-17593; [10/N] Remove resolved regular expressions when unsubscribed

### PR DESCRIPTION
**Pull request by dajac**
This patch does a few things:
1) It cleans up resolved regular expressions when they are unsubscribed from. It covers the regular leave/fenced paths for the new protocol and it also covers the LeaveGroup API as new members could be removed via the admin API.
2) It ensures that tombstones for  resolved regular expressions are generated on the conversion patch from consumer to classic group.
3) It fixes [KAFKA-18116](https://issues.apache.org/jira/browse/KAFKA-18116) because I faced the same issue while working on the LeaveGroup API. It adds an integration test for this case too.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java [hunk 1]
```diff
@@ -6128,11 +6183,11 @@ private CoordinatorResult<LeaveGroupResponseData, CoordinatorRecord> classicGrou
                     // in which case we expect the MemberId to be undefined.
                     if (!UNKNOWN_MEMBER_ID.equals(memberId)) {
                         throwIfInstanceIdIsFenced(member, groupId, memberId, instanceId);
+                        throwIfMemberDoesNotUseClassicProtocol(member);
```
Comment (dajac) [inline group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java:6000]: This is the fix for https://issues.apache.org/jira/browse/KAFKA-18116. @dongnuo123 What do you think about it?

File: group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java [hunk 2]
```diff
@@ -3224,6 +3228,59 @@ private void replaceMember(
...
 
+    /**
+     * Maybe delete the resolved regular expression associated to the provided member if
+     * it was the last subscribed member to it.
+     *
+     * @param records   The record accumulator.
+     * @param group     The group.
+     * @param member    The member removed from the group.
+     */
+    private void maybeDeleteResolvedRegularExpression(
```
Comment (jeffkbkim) [inline group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java:3053]: can we share implementations in these two methods?

File: group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java (hunk 1 above)
Comment (dongnuo123) [reply group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java:6000]: Should we check `validateOnlineDowngradeWithFencedMember` if a consumer member leaves the group though admin client?

File: group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java [hunk 3]
```diff
@@ -3224,6 +3228,59 @@ private void replaceMember(
...
+            records.add(newConsumerGroupRegularExpressionTombstone(
+                group.groupId(),
+                member.subscribedTopicRegex()
+            ));
+        }
+    }
+
+    /**
+     * Maybe delete the resolved regular expressions associated with the provided members
+     * if they were the last ones susbcribed to them.
```
Comment (dongnuo123) [inline group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java]: nit: subscribed

File: group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java (hunk 1 above)
Comment (dajac) [reply group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java:6000]: I suppose that we should. However, as it is orthogonal to this PR, I suggest to tackle it separately. Could you please file a jira for it?

File: group-coordinator/src/main/java/org/apache/kafka/coordinator/group/modern/consumer/ConsumerGroup.java [hunk 4]
```diff
@@ -655,24 +722,28 @@ public void createGroupTombstoneRecordsWithReplacedMember(
...
         });
 
-        records.add(GroupCoordinatorRecordHelpers.newConsumerGroupSubscriptionMetadataTombstoneRecord(groupId()));
-        records.add(GroupCoordinatorRecordHelpers.newConsumerGroupEpochTombstoneRecord(groupId()));
+        resolvedRegularExpressions.keySet().forEach(regex ->
+            records.add(GroupCoordinatorRecordHelpers.newConsumerGroupRegularExpressionTombstone(groupId, regex))
+        );
+
+        records.add(GroupCoordinatorRecordHelpers.newConsumerGroupSubscriptionMetadataTombstoneRecord(groupId));
+        records.add(GroupCoordinatorRecordHelpers.newConsumerGroupEpochTombstoneRecord(groupId));
```
Comment (jeffkbkim) [inline group-coordinator/src/main/java/org/apache/kafka/coordinator/group/modern/consumer/ConsumerGroup.java:746]: i recall we no longer should care about ordering of record types. is this still the case?

File: group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java [hunk 5]
```diff
@@ -3038,6 +3042,59 @@ private void replaceMember(
...
+    /**
+     * Maybe delete the resolved regular expressions associated with the provided members
+     * if they were the last ones subscribed to them.
+     *
+     * @param records   The record accumulator.
+     * @param group     The group.
+     * @param members   The member removed from the group.
+     * @return The set of deleted regular expressions.
+     */
+    private Set<String> maybeDeleteResolvedRegularExpressions(
```
Comment (jeffkbkim) [inline group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java:3075]: should we add a test for this? or include in another test where a single member is subscribed to topic "foo" as both a topic name and as a regex, i.e. "foo" and "foo*"

File: core/src/test/scala/unit/kafka/server/GroupCoordinatorBaseRequestTest.scala [hunk 6]
```diff
@@ -660,6 +660,24 @@ class GroupCoordinatorBaseRequestTest(cluster: ClusterInstance) {
     )
   }
 
+  protected def classicLeaveGroup(
```
Comment (jeffkbkim) [inline core/src/test/scala/unit/kafka/server/GroupCoordinatorBaseRequestTest.scala:663]: what's the difference between this and leaveGroupWithOldProtocol except for the expected response?

**Review by jeffkbkim — COMMENTED** [review]
Thanks for the PR. left some comments

File: group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java (hunk 1 above)
Comment (dongnuo123) [reply group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java:6000]: https://issues.apache.org/jira/browse/KAFKA-18150

File: group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java (hunk 2 above)
Comment (dajac) [reply group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java:3053]: i considered calling `maybeDeleteResolvedRegularExpressions` from this one but i found that building a Map and a Set was not necessary so i opted for keeping a slightly more efficient implementation here.

File: group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java (hunk 5 above)
Comment (dajac) [reply group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java:3075]: this is tested by `testResolvedRegularExpressionsRemovedWhenConsumerMembersRemovedByAdminApi`. i am not sure to follow your suggestion about having a member subscribed to both `foo` and `foo*`. this code does not look at topic names at all. could you please elaborate?

File: group-coordinator/src/main/java/org/apache/kafka/coordinator/group/modern/consumer/ConsumerGroup.java (hunk 4 above)
Comment (dajac) [reply group-coordinator/src/main/java/org/apache/kafka/coordinator/group/modern/consumer/ConsumerGroup.java:746]: we still follow the convention when we generate them but the ordering does not matter on replay.

File: core/src/test/scala/unit/kafka/server/GroupCoordinatorBaseRequestTest.scala (hunk 6 above)
Comment (dajac) [reply core/src/test/scala/unit/kafka/server/GroupCoordinatorBaseRequestTest.scala:663]: `leaveGroupWithOldProtocol` assumes that the memberid sent must be the one received. when the admin api is used to remove a static member, the memberid is not set but the correct one is returned. this is why i put a raw method.

File: group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java (hunk 5 above)
Comment (jeffkbkim) [reply group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java:3075]: I might have misplaced the comment. I meant in `ConsumerGroup#computeSubscribedTopicNames`, it is possible to double decrement `subscribedTopicsNames` once from subscribed topic name and once from subscribed regex.

Maybe we can add it in testComputeSubscribedTopicNames?

File: core/src/test/scala/unit/kafka/server/GroupCoordinatorBaseRequestTest.scala (hunk 6 above)
Comment (jeffkbkim) [reply core/src/test/scala/unit/kafka/server/GroupCoordinatorBaseRequestTest.scala:663]: can't we call `classicLeaveGroup` inside `leaveGroupWithOldProtocol` to reuse the code?

File: group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java (hunk 5 above)
Comment (dajac) [reply group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java:3075]: ah, got it. let me extend that test.

File: core/src/test/scala/unit/kafka/server/GroupCoordinatorBaseRequestTest.scala (hunk 6 above)
Comment (dajac) [reply core/src/test/scala/unit/kafka/server/GroupCoordinatorBaseRequestTest.scala:663]: yes, we can. let me update that too.

**Review by jeffkbkim — APPROVED** [review]
LGTM once tests pass. thanks!

**Pull request merged**